mopro-ffi/src/lib.rs

```diff
@@ -66,32 +111,77 @@ impl From<MoproError> for FFIError {
     }
 }
 
-// Test functions (TODO - consider removing)
-
-fn add(a: u32, b: u32) -> u32 {
-    a + b
+#[derive(Debug, Clone, Default)]
+pub struct G1 {
+    pub x: String,
+    pub y: String,
 }
 
-fn hello() -> String {
-    "Hello World from Rust".to_string()
+#[derive(Debug, Clone, Default)]
+pub struct G2 {
+    pub x: Vec<String>,
+    pub y: Vec<String>,
 }
 
-// TODO: Remove me
-// UniFFI expects String type
-// See https://mozilla.github.io/uniffi-rs/udl/builtin_types.html
-// fn run_example(wasm_path: String, r1cs_path: String) -> Result<(), MoproError> {
-//     circom::run_example(wasm_path.as_str(), r1cs_path.as_str())
-// }
-
-uniffi::include_scaffolding!("mopro");
-
-#[cfg(test)]
-mod tests {
-    use super::*;
+#[derive(Debug, Clone, Default)]
+pub struct ProofCalldata {
+    pub a: G1,
+    pub b: G2,
+    pub c: G1,
+}
 
-    #[test]
-    fn add_works() {
-        let result = add(2, 2);
-        assert_eq!(result, 4);
-    }
+// This macro should be used in dependent crates
+//
+// This macro handles getting relevant functions into
+// scope and calling uniffi
+//
+// There should be a user defined `zkey_witness_map` function
+// that maps zkey file stub to a witness generation function
+// see test-e2e/src/lib.rs for an example
+#[macro_export]
+macro_rules! app {
+    () => {
+        use mopro_ffi::{GenerateProofResult, MoproError, ProofCalldata, G1, G2};
+        use std::collections::HashMap;
+
+        fn generate_halo2_proof(
+            in0: HashMap<String, Vec<String>>,
+        ) -> Result<GenerateProofResult, MoproError> {
+            mopro_ffi::generate_halo2_proof(in0)
+        }
+
+        fn verify_halo2_proof(in0: Vec<u8>, in1: Vec<u8>) -> Result<bool, MoproError> {
+            mopro_ffi::verify_halo2_proof(in0, in1)
+        }
+
+        fn generate_circom_proof(
+            in0: String,
+            in1: HashMap<String, Vec<String>>,
+        ) -> Result<GenerateProofResult, MoproError> {
+            let name = std::path::Path::new(in0.as_str()).file_name().unwrap();
+            if let Ok(witness_fn) = zkey_witness_map(&name.to_str().unwrap()) {
+                mopro_ffi::generate_circom_proof_wtns(in0, in1, witness_fn)
+            } else {
+                Err(MoproError::CircomError("Unknown ZKEY".to_string()))
+            }
+        }
```

Comment:
not very sure how it works
will it load both circom and halo2 provers?
I think normal users don't use both circom and halo2 provers

Reply:
It won't. If the halo2 feature is not enabled the function will simply throw. We need to define all the functions in order for the udl to compile though.